api: implement tag immutability (PROJQUAY-1253)

dmesser opened the pull request:
This implements support for immutable tags, including support in both old and new UIs as well as corresponding log events and unit test. Immutable tags have the following characteristics:

* only users with write permissions to the repository can set tags to be immutable
 * only users with admin permissions to the repository can make a tag mutable again
 * immutable tags cannot be overwritten
 * immutable tags cannot be deleted
 * manifests which have immutable tags pointing to them cannot be deleted
 * manifests with immutable tags cannot be deleted
 * immutable tags cannot be restored to via time machine (reversion)
 * deletion of repositories containing immutable tags (or orgs containing such repos) is no impacted by this
 * immutable tags cannot expire
   * one cannot set an expiry date on an immutable tag
   * one cannot make a tag immutable if it has an expiry date
 * manifests with immutable tags cannot have their labels changed (add or removed)
 * new tags can be created off immutable tags
 * tags mutability can be toggled via the API
 * tags can be made immutable by pushing a manifest with label (via `Dockerfile` / `Containerfile` called `quay.immutable=true`
   * adding the label via mutable labels in Quay, will not set the the immutability flag as it is a reserved label for push operations
   * the label cannot be removed as it is a built-in read-only label
   * this label cannot coexist with the `quay.expires-after` label, it will be ignored
 * tag mutability settings via the API take precedence over labels
   * tags without a `quay.immutable` label can be set to immutable via the API
   * tags with the label `quay.immutable=true` can be set mutable via the API
 * repositories with immutable tags cannot be converted to a mirror as the mirroring could fail to overwrite immutable tags
 * orgs with repositories that have immutable tags cannot be converted to a cache org as the caching could fail to overwrite immutable tags
 * tag immutability is only implemented for Docker v2s2/OCI specifications and formats

<img width="868" alt="Screenshot 2023-05-31 at 16 28 13" src="https://github.com/quay/quay/assets/12664117/85eb36cd-dc17-4702-a7db-af219365e98e">
<img width="1302" alt="Screenshot 2023-05-31 at 16 27 32" src="https://github.com/quay/quay/assets/12664117/2aa7db65-8a4c-4518-bf7c-7fadb27fd7b4">
<img width="1307" alt="Screenshot 2023-05-31 at 16 27 27" src="https://github.com/quay/quay/assets/12664117/9c7a3af2-b521-438d-acd0-7076bbfa3054">
<img width="1307" alt="Screenshot 2023-05-31 at 16 27 11" src="https://github.com/quay/quay/assets/12664117/db115bbf-ac2b-4d5f-94e4-adc5cb933ca7">
<img width="1315" alt="Screenshot 2023-06-03 at 21 25 24" src="https://github.com/quay/quay/assets/12664117/c57e198c-02d9-4c71-8d28-29f53ca44a44">
<img width="1313" alt="Screenshot 2023-06-03 at 21 22 00" src="https://github.com/quay/quay/assets/12664117/4a70f1ee-1eba-4fce-8187-2134c9ea47a4">
<img width="1319" alt="Screenshot 2023-06-03 at 21 21 38" src="https://github.com/quay/quay/assets/12664117/c61cb4f4-7fef-430f-9047-29c32d4971d3">
<img width="1321" alt="Screenshot 2023-06-03 at 00 25 47" src="https://github.com/quay/quay/assets/12664117/f4612602-2ef9-45b4-801a-ecf4169649e9">
<img width="1322" alt="Screenshot 2023-06-03 at 00 25 27" src="https://github.com/quay/quay/assets/12664117/2803ea43-bb1e-4442-8a9b-486e82651388">
<img width="1321" alt="Screenshot 2023-06-03 at 00 25 17" src="https://github.com/quay/quay/assets/12664117/386e6a7a-4774-459d-8ab3-debdeb780db2">
<img width="1023" alt="Screenshot 2023-05-31 at 16 28 53" src="https://github.com/quay/quay/assets/12664117/b94741de-28bb-4823-92d6-cf5ce76fe6f1">

bcaton85 commented on web/src/resources/TagResource.ts:
```diff
@@ -266,62 +279,112 @@ export async function bulkDeleteTags(
+    }
+    throw bulkDeleteError;
+  }
+}
```
You could probably use [throwIfError](https://github.com/quay/quay/blob/master/web/src/resources/ErrorHandling.ts#L31) here

dmesser replied:
Cool, TIL. Updated.,

bcaton85 commented on web/src/routes/RepositoryDetails/Tags/TagsActions.tsx:
```diff
@@ -57,8 +65,29 @@ export default function TagActions(props: TagActionsProps) {
+  dropdownItems.push(
     <DropdownItem
       key="delete-tag-action"
+      isDisabled={props.tags.some((tag: Tag) => tag.immutable)}
```
This should be added to the edit labels and edit expiration options too

dmesser replied:
Great catch, updated.

bcaton85 reviewed (commented):
Great work. Gave it a first look through. Still testing it out locally. A concern I have is this isn't behind a feature flag and includes a not-so-simple DB migration, so reverting won't be possible. What would be the approach of rolling back if this causes issues?

dmesser commented on data/model/oci/tag.py:
```diff
@@ -459,6 +500,42 @@ def retarget_tag(
+    """
+    Sets the immutable flag on the given tag.
+    """
+    with db_transaction():
```
Got it.

dmesser commented on data/model/repository.py:
```diff
@@ -643,7 +629,27 @@ def get_repository_state(namespace_name, repository_name):
+    )
+
+
+def set_repository_state(repo, state, raise_on_error=False):
```
Updated

dmesser commented on data/migrations/versions/5b8dc452f5c3_add_tag_immutability_column_and_logentry_kinds.py:
```diff
@@ -0,0 +1,70 @@
+            nullable=False,
+            server_default=sa.sql.expression.false(),
+        ),
+    )
```
I am not sure about other databases, but Postgres should be quite fast if I understand correctly (since it's a non-volatile default value).  Given that we deprecated MySQL for a while now, the advantage of avoiding the overhead of constantly doing joins and lookups for the majority of the users outweighs a slow migration / Quay update for a very small number of users.
On the other hand, how heavy would the overhead be if a join from a relatively small and likely completely cached table constantly calculates the tag immutability info?

bcaton85 replied:
if we resorted to using the joins the overhead would be low but ideally we could just add the column and prevent the database structure from becoming fragmented into different tables. To do that IMO we need some sort of confirmation that when the migration runs on Quay.io it doesn't take more resources than expected and impact the service. I'm open to other opinions @syed @kleesc

dmesser replied:
My hope is that adding a column that has a static default value is almost a no-op for Postgres and it's MVCC, because the value can be computed when read and doesn't have to be written to a new row internally unless actually set to `True` or due to another update of the row (e.g. expiry date)

dmesser commented on data/model/oci/tag.py:
```diff
@@ -459,6 +500,42 @@ def retarget_tag(
+    return _set_tag_immutability(repository_id, tag, immutable)
+
+
+def _set_tag_immutability(repository_id, tag, immutable):
```
updated

dmesser commented on data/model/oci/tag.py:
```diff
@@ -770,6 +893,7 @@ def fetch_paginated_autoprune_repo_tags_by_number(repo_id, max_tags_allowed: int
                 Tag.repository_id == repo_id,
                 (Tag.lifetime_end_ms >> None) | (Tag.lifetime_end_ms > now_ms),
                 Tag.hidden == False,
+                Tag.immutable == False,
```
I updated the implementation to calculate the offset according to the total number of immutable tags in the repo. The expectation is that immutable tags can never be auto-pruned; hence, the total number of tags after pruning might be higher than the pruning configuration.

bcaton85 reviewed (commented):
Tested the changes out locally, works well, great job. There's a couple of smaller issues that you could address if you wanted to but i don't think are required to be merged:
- Selecting multiple immutable tags and setting them to mutable doesn't work in the old ui, no error, just reload the page and the tags are still immutable.
- If pushing an image with both expires and immutable labels, expires overrides immutable and immutable is never set
- Pushing to an immutable tag causes client retries, podman retried 5 pushes before failing due to immutable tag. I believe the client isn't aware how to handle a 412 response so it just does a general retry.

This PR should also be broken down into multiple PR's. (db migration, API, UI, setting via label). It's difficult to review 3000+ changes. Also if the code does cause an issue we cannot revert because of the DB migration and we don't have a feature flag to turn it off so we would be forced to fix it forward.

Thank you for working on this.

dmesser commented on data/model/organization.py:
```diff
@@ -194,3 +195,19 @@ def add_user_as_admin(user_obj, org_obj):
+    """
+    return (
+        Tag.select()
+        .join(Repository)
```
To find immutable tags inside an organization I have to traverse through the repository, since Tags are linked to Repositories and only those are linked to Users/Organizations. I don't need to explictly use the Repository relation elsewhere in the query since peewee will automatically figure out through the foreign key relationships which fields to use for the join

dmesser commented on web/src/routes/RepositoryDetails/Tags/TagsTable.tsx:
```diff
@@ -256,6 +263,15 @@ export default function TagsTable(props: TableProps) {
+              <Tooltip content="Tag Immutability">
+                <LockOpenIcon />
+              </Tooltip>
+            </Th>
```
adjusted for a more generic status column in which we later can also capture other tag state info, like whether it has certain referrer types (e.g. signature or SBOM)

dmesser commented:
I've change the "immutable" column in the tags table to a more generic status column which now tracks the immutability via a label. Mutable tags won't carry a label.

<img width="1273" alt="Screenshot 2024-05-19 at 21 41 17" src="https://github.com/quay/quay/assets/12664117/f2c2038c-b373-4a2d-a5a0-deb2dd1c06ca">

dmesser commented on endpoints/api/tag.py:
```diff
@@ -60,6 +66,89 @@ def _tag_dict(tag):
 
 
+def _set_tag_immutable(tag_ref, repository, namespace):
+    if not AdministerRepositoryPermission(namespace, repository).can():
```
Initially I thought that only admins should be able to control mutability but rethinking it now its probably too restrictive.

dmesser commented:
@bcaton85 Thank you very much for your thorough review. I addressed all the feedback, with the exception of that seemingly unused join (it is actually used).

Regarding your observations:

> Selecting multiple immutable tags and setting them to mutable doesn't work in the old ui, no error, just reload the page and the tags are still immutable.

I couldn't reproduce this on my end. One thing that might cause requests to be stalled is too many background requests.

> If pushing an image with both expires and immutable labels, expires overrides immutable and immutable is never set

That is correct and intentional for now, albeit there is user feedback that people want to mark tags immutable but still give them an expiry label, which I plan to address via an opt-in in a follow-up PR.

> Pushing to an immutable tag causes client retries, podman retried 5 pushes before failing due to immutable tag. I believe the client isn't aware how to handle a 412 response so it just does a general retry.

I didn't see this via `skopeo` or `docker` , it seems this got added to podman recently: https://github.com/containers/podman/commit/c3c0c4ab96e43f8364756971f34387289b3b50b2
While I like the retry I don't think it should retry on 413, but I think this is a cosmetic error.

Please let me know if I should now break up the PR in various smaller pieces for it got get merged.
